ujt/callbacks/signal_callbacks.py

```diff
@@ -0,0 +1,77 @@
+        id_constants.SIGNAL_TAG_DELETE,
+        id_constants.SIGNAL_TAG_SAVE,
+    ),
+    id_constants.SIGNAL_APPLIED_TAG_UPDATE: (
+        id_constants.SIGNAL_APPLIED_TAG_ADD,
+        id_constants.SIGNAL_APPLIED_TAG_REMOVE,
+        id_constants.SIGNAL_APPLIED_TAG_MODIFY,
+    ),
+    id_constants.SIGNAL_VIEW_UPDATE: (
+        id_constants.SIGNAL_VIEW_CREATE,
+        id_constants.SIGNAL_VIEW_DELETE,
+        id_constants.SIGNAL_VIEW_MODIFY,
+    ),
+    id_constants.SIGNAL_STYLE_UPDATE: (
+        id_constants.SIGNAL_STYLE_SAVE,
+        id_constants.SIGNAL_STYLE_DELETE,
+    ),
+    id_constants.SIGNAL_COMPOSITE_TAGGING_UPDATE: (
+        id_constants.SIGNAL_TAG_UPDATE,
+        id_constants.SIGNAL_APPLIED_TAG_UPDATE,
+        id_constants.SIGNAL_VIEW_UPDATE,
+        id_constants.SIGNAL_STYLE_UPDATE,
+    ),
+}
+
+
+def generate_generic_update_signal(*args, **kwargs):
+    """Combines individual signals into a composite update signal.
+
+    This allows other callbacks to avoid having to register each individual type of signal
+    (create, remove, etc...) when they are interested in any change to the underlying data structure.
+    However, callbacks have the flexibility to only listen to the signals that correspond to
+    specific types of operations.
+
+    For example, consider the style functionality.
+    The save style button callback needs to update the properties of the save toast,
+    and saves the style as a side effect. This callback updates the SIGNAL_STYLE_SAVE signal.
+    The load and delete callback needs to update the properties of the text fields,
+    and deletes the style as a side effect. This callback updates the SIGNAL_STYLE_DELETE signal.
+
+    The SIGNAL_STYLE_UPDATE is used to trigger the update_cytoscape_stylesheet callback.
+    If we want to update the SIGNAL_STYLE_UPDATE directly from the save and delete buttons,
+    we have to combine the two callbacks to saving and deleting (since Dash only supports assigning
+    to an output with a single callback).
+    In order to avoid combining the callbacks, we let them each produce their own update signals
+    and combine them with this callback is a workaround.
+
+    Returns:
+        A OK signal constant.
+    """
+    return constants.OK_SIGNAL
+
+
+def generate_composite_signals():
+    for output_signal_id, input_signal_ids in COMPOSITE_SIGNAL_MAP.items():
+        dash_output = Output(output_signal_id, "children")
+        dash_inputs = [
+            Input(input_signal_id, "children") for input_signal_id in input_signal_ids
+        ]
+        # Usually we use app.callback as a decorator, which is syntatic sugar
+        # app.callback(outputs=..., inputs=..., state=..., prevent_initial_call=...)(decorated_function)
+        # Here we call the decorator directly to programatically register the callbacks.
+        app.callback(dash_output, dash_inputs, prevent_initial_call=True)(
+            generate_generic_update_signal
+        )
```

Comment:
The only major change is made here. We previously had individual functions that associated input and output signals, and returned a constant OK signal. Now, we can register these callbacks dynamically and eliminate some duplicated code.